### PR TITLE
GH-747: Review 2

### PR DESCRIPTION
**Pull request by bertllll**

### CODE REVIEW

File: node/src/accountant/db_access_objects/receivable_dao.rs [hunk 1]
```diff
@@ -38,11 +37,11 @@ pub enum ReceivableDaoError {
...
-impl From<PersistentConfigError> for ReceivableDaoError {
-    fn from(input: PersistentConfigError) -> Self {
-        ReceivableDaoError::ConfigurationError(format!("{:?}", input))
-    }
-}
+// impl From<PersistentConfigError> for ReceivableDaoError {
+//     fn from(input: PersistentConfigError) -> Self {
+//         ReceivableDaoError::ConfigurationError(format!("{:?}", input))
+//     }
+// }
```
Comment (utkarshg6) [inline node/src/accountant/db_access_objects/receivable_dao.rs:44]: I think you'd rather prefer to delete these lines instead of commenting.

File: node/src/accountant/db_access_objects/receivable_dao.rs [hunk 2]
```diff
@@ -362,37 +372,40 @@ impl ReceivableDaoReal {
...
+                "Update for received payment with {amount} wei ran without producing a change in \
+                the database. At the same time, it's been found that wallet address {address} refers \
+                to {x} records while it always should have none or one."
+            ))),
+        };
+        info!(
+            logger,
+            "Received {amount} wei from unknown debtor {address} in the block {block_number}."
+        );
+        Ok(())
```
Comment (utkarshg6) [inline node/src/accountant/db_access_objects/receivable_dao.rs:401]: If you'd like you can move this `info` log to the match arm with the `0` and also return the `Ok(())` there.

File: node/src/accountant/db_access_objects/receivable_dao.rs [hunk 3]
```diff
@@ -442,16 +455,13 @@ impl ReceivableDaoReal {
...
         logger: &Logger,
         payments: &[BlockchainTransaction],
     ) {
         fn finalize_report(mut report_lines: Vec<String>, sum_of_all_txs: u128) -> String {
-            report_lines.push(
-                format!("{:10} {:42} {:<18}", "TOTAL", "", sum_of_all_txs),
-            );
-            report_lines
-            .join("\n")
+            report_lines.push(format!("{:10} {:42} {:<18}", "TOTAL", "", sum_of_all_txs));
```
Comment (utkarshg6) [inline node/src/accountant/db_access_objects/receivable_dao.rs:463]: I don't know why are you keeping this extra padding of `{:42}`. I believe, either you should combine them like this:

```rust
report_lines.push(format!("{:52} {:<18}", "TOTAL", sum_of_all_txs));
```

or maybe just remove it if it's unnecessary. Maybe you've a good reason to do it, if that's the case, you can keep it.

**Review by utkarshg6 — COMMENTED** [review]
Continue at node/src/accountant/db_access_objects/receivable_dao.rs:1036

File: node/src/database/test_utils/transaction_wrapper_mock.rs [hunk 4]
```diff
@@ -66,329 +118,226 @@ impl TransactionWrapper for TransactionWrapperMock {
...
+            conn,
+            queue_of_statements: RefCell::new(altered_stmts_queue),
+        };
+
+        Self {
+            setup: Either::Left(setup),
+        }
+    }
+
+    pub fn new_with_prod_code_and_altered_stmts(
```
Comment (utkarshg6) [inline node/src/database/test_utils/transaction_wrapper_mock.rs:259]: Similar to the above comment we can rename this function name to `construct_with_prod_code_and_altered_stmts`.

File: node/src/accountant/db_access_objects/receivable_dao.rs [hunk 5]
```diff
@@ -1030,24 +1026,25 @@ mod tests {
...
-        let mocked_transaction = Box::new(
-            TransactionWrapperMock::default()
-                .prepare_params(&prepare_params_arc)
-                .prepare_results(prepare_results),
+            Some(panic_causing_conn),
+            vec![
+                None,
+                None,
+                None,
+                Some(AlteredStmtByOrigin::IdenticalWithProdCode),
```
Comment (utkarshg6) [inline node/src/accountant/db_access_objects/receivable_dao.rs:1041]: Can you please add a comment beside a `None` and the `Some`, explaining what they are.

File: node/src/accountant/db_access_objects/receivable_dao.rs [hunk 6]
```diff
@@ -1152,26 +1157,61 @@ mod tests {
...
+        let txn_inner_builder =
+            TransactionInnerWrapperMockBuilder::default().prepare_results(results);
+        let txn = TransactionSafeWrapper::new_with_builder(txn_inner_builder);
+        let suspect = BlockchainTransaction {
+            block_number: 1234,
+            from: wallet,
+            wei_amount: 1_000_000_000,
+        };
+
+        let logger = Logger::new(test_name);
```
Comment (utkarshg6) [inline node/src/accountant/db_access_objects/receivable_dao.rs:1206]: Can you please remove the line above this one.

File: node/src/accountant/db_access_objects/receivable_dao.rs [hunk 7]
```diff
@@ -1152,26 +1157,61 @@ mod tests {
...
-                expected_panic_msg.to_string()
-            ))
-        )
+        without producing a change in the database, despite the record for wallet \
+        0x00000000000000000000000000000000626c6168 exists.";
+        assert_eq!(result, Err(RusqliteError(expected_panic_msg.to_string())))
+    }
+
+    #[test]
+    fn verify_possibly_unknown_wallet_returns_error_finding_duplicated_records_for_single_wallet() {
```
Comment (dnwiebe) [inline node/src/accountant/db_access_objects/receivable_dao.rs:1172]: This test tests an impossible condition: that there are several records in the table with the same wallet address.

The actual definition for this table, from `db_initializer.rs`, is:
```
create table if not exists receivable (
        wallet_address text primary key,
        balance_high_b integer not null,
        balance_low_b integer not null,
        last_received_timestamp integer not null
) strict
```
Notice that the `wallet_address` column is a primary key; so uniqueness will be enforced. Also notice that the balance column is actually two columns, both of which are `not null`, and there's a `last_received_timestamp` column that's also `not null` which doesn't appear in the test table. I don't know whether there's an important difference between the `text` and `string` data types.

It's important to test with the same table schema you have in production. It might even be worth putting the table definitions in manifest constants and using the same constant in both places.

File: node/src/accountant/db_big_integer/big_int_db_processor.rs [hunk 8]
```diff
@@ -13,39 +13,39 @@ use std::iter::once;
...
+        conn: Either<&dyn ConnectionWrapper, &TransactionSafeWrapper>,
         config: BigIntSqlConfig<'params, T>,
     ) -> Result<(), BigIntDatabaseError> {
         let main_sql = config.main_sql;
         let mut stm = Self::prepare_statement(conn, main_sql);
         let params = config
             .params
             .merge_other_and_wei_params((&config.params.wei_change_params).into());
-        match stm.execute(params.as_slice()) {
+        match stm.execute(<Vec<RusqliteNativePrimitivePair>>::from_iter(params).as_slice()) {
```
Comment (dnwiebe) [inline node/src/accountant/db_big_integer/big_int_db_processor.rs:46]: What's the purpose of this type change?

File: node/src/accountant/db_big_integer/big_int_db_processor.rs [hunk 9]
```diff
@@ -277,35 +280,56 @@ impl<'params, T: TableNameDAO> BigIntSqlConfig<'params, T> {
...
-type DisplayableRusqliteParamPair<'params> = (&'params str, &'params dyn DisplayableParamValue);
+// (<param_name_str>, <param_value_behind_reference>)
+pub struct RusqliteParamPair<'params> {
+    pub name: &'params str,
+    pub value: &'params dyn ToSql,
+}
 
-// To be able to display things that implement ToSql
-pub trait DisplayableParamValue: ToSql + Display {}
+type RusqliteNativePrimitivePair<'params> = (&'params str, &'params dyn ToSql);
```
Comment (utkarshg6) [inline node/src/accountant/db_big_integer/big_int_db_processor.rs:289]: What does the first part of the tuple represent? Is it a data value? If it's not, can you please make it more intention revealing?

File: node/src/accountant/db_big_integer/big_int_db_processor.rs [hunk 10]
```diff
@@ -315,78 +339,96 @@ impl<'params> SQLParamsBuilder<'params> {
...
             },
-            params_except_wei_change: params,
+            other_than_wei_change_params: params,
         }
     }
 
-    fn expand_wei_params(wei_change_spec: WeiChange) -> ((String, String), (i64, i64)) {
+    fn expand_wei_params(
+        wei_change_spec: WeiChange,
+    ) -> (RusqliteNumParamFormNamed, RusqliteNumParamFormNamed) {
```
Comment (utkarshg6) [inline node/src/accountant/db_big_integer/big_int_db_processor.rs:377]: I'd like if we can directly infer what the first and the second element mean. It would be better if we have this struct as the return type, something like:

```rust
// Maybe use this 
WeiChangeAsHighAndLowBytes {
    high_bytes: RusqliteNumParamFormNamed,
    low_bytes: RusqliteNumParamFormNamed,
}
```

File: node/src/accountant/db_big_integer/big_int_db_processor.rs [hunk 11]
```diff
@@ -419,53 +461,75 @@ impl Display for ByteMagnitude {
...
+pub enum ParamByUse<'params> {
+    BeforeAndAfterOverflow {
+        name: &'params str,
+        value: &'params dyn DisplayableParamValue,
+    },
+    BeforeOverflowOnly {
+        name: &'params str,
+        value: &'params dyn DisplayableParamValue,
+    },
 }
```
Comment (utkarshg6) [inline node/src/accountant/db_big_integer/big_int_db_processor.rs:496]: If you'd like you can introduce this new struct to eliminate code duplication.

```rust
struct OverflowAdjuster {
        name: &'params str,
        value: &'params dyn DisplayableParamValue,
}

pub enum ParamByUse<'params> {
    BeforeAndAfterOverflow(OverflowAdjuster),
    BeforeOverflowOnly(OverflowAdjuster),
}
```

File: node/src/accountant/db_big_integer/big_int_db_processor.rs [hunk 12]
```diff
@@ -506,10 +570,16 @@ pub trait TableNameDAO: Debug + Send {
...
+        amount_to_add: u128,
+    },
     //This means that the supplied amount is internally converted into a signed number and
     //so even though the related SQL contains + signs the resulting math operation is subtraction
-    Subtraction(&'static str, u128),
+    Subtraction {
+        name_without_suffix: &'static str,
+        amount_to_subtract: u128,
+    },
 }
```
Comment (utkarshg6) [inline node/src/accountant/db_big_integer/big_int_db_processor.rs:583]: I think we should change it too to eliminate code duplication:

```rust
struct WeiChange {
        name_without_suffix: &'static str,
        amount: u128,
        operation: WeiChangeOperation // Addition or Subtraction
}

enum WeiChangeOperation {
   Addition,
   Subtraction
}
```

**Review by utkarshg6 — COMMENTED** [review]
Nicely done

File: node/src/database/rusqlite_wrappers.rs [hunk 13]
```diff
@@ -1,24 +1,62 @@
...
+// The actual blocker has always rooted in the hardly graspable relationships of serialized
+// lifetimes affecting each other. The choices made beyond the third-party library can be found
+// reasonable, like those ensuring that the database connection must stay alive as long as there
+// are any active 'Transactions' or 'Statements'.
+//
+// The use of an explicit, object-like transaction from 'rusqlite' is rare in our code. That's
+// probably why we'd managed to live so long without feeling the need for this new mock. While we
+// did write some code using it, we always acknowledged we had to leave that piece of logic untested.
+// The actual problem with a mocked transaction has been that we usually need to combine it with
+// an also mocked connection. The transaction is springs from the connection. The need to put
```
Comment (utkarshg6) [inline node/src/database/rusqlite_wrappers.rs:31]: You can remove `is`, it's something you mentioned while you were sitting beside me.

File: node/src/database/rusqlite_wrappers.rs [hunk 14]
```diff
@@ -1,24 +1,62 @@
...
+// this object, strictly depending on a reference, into another mock makes us face a nontrivial
```
Comment (utkarshg6) [inline node/src/database/rusqlite_wrappers.rs:32]: Instead of "this object", we can say, "this mock"..

File: node/src/database/rusqlite_wrappers.rs [hunk 15]
```diff
@@ -30,10 +68,8 @@ impl ConnectionWrapper for ConnectionWrapperReal {
     fn prepare(&self, query: &str) -> Result<Statement, Error> {
         self.conn.prepare(query)
     }
-    fn transaction<'a>(&'a mut self) -> Result<Box<dyn TransactionWrapper + 'a>, Error> {
-        self.conn
-            .transaction()
-            .map(|tx| Box::new(TransactionWrapperReal::new(tx)) as Box<dyn TransactionWrapper + 'a>)
+    fn transaction(&mut self) -> Result<TransactionSafeWrapper, Error> {
+        self.conn.transaction().map(TransactionSafeWrapper::new)
```
Comment (utkarshg6) [inline node/src/database/rusqlite_wrappers.rs:72]: Thanks for removing the lifetime from here. :tada:

File: node/src/database/rusqlite_wrappers.rs [hunk 16]
```diff
@@ -43,44 +79,102 @@ impl ConnectionWrapperReal {
     }
 }
 
-pub trait TransactionWrapper: Debug {
+// Whole point of this outer wrapper that is common in both the real and mock transactions is to
+// make a chance to deconstruct this whole code structure in place. It plays a crucial role in
+// the call of 'commit'. Note that a normally written mock with the direct use of a trait object
+// doesn't allow to be consumed by any of its methods because of the rules declared for trait
+// objects in Rust. They say clearly that we can never access one otherwise than via a reference.
```
Comment (utkarshg6) [inline node/src/database/rusqlite_wrappers.rs:86]: They say clearly that we can access it via `&self`, `&mut self` and not `self`.

Reason: This is more explicit.

File: node/src/database/rusqlite_wrappers.rs [hunk 17]
```diff
@@ -43,44 +79,102 @@ impl ConnectionWrapperReal {
...
+// However, to have a thing be consumed by itself we need to be possessing the full ownership.
+//
+// Leaving an already enclosed, committed, transaction around would expose us to a risk. Let's
+// imagine somebody trying to use it the second time, while the actual element providing
+// the connection has been taken away and deconstructed, then this wrapper could still live on,
+// having noting to use inside.
+//
+// Second, caution is much desirable here because the wrapper hides a potential, hard-to-debug,
```
Comment (utkarshg6) [inline node/src/database/rusqlite_wrappers.rs:94]: Let's mention that this is not a production code problem and is just an issue during writing tests.

File: node/src/database/rusqlite_wrappers.rs [hunk 18]
```diff
@@ -43,44 +79,102 @@ impl ConnectionWrapperReal {
...
+// segmentation error that might the OS might raise if we fail to be early deconstructing
+// the transaction we created and put inside. It has an unsafe reference pointing back in to
+// the same object, to a field that hosts its own, extra, database connection that it can freely
+// carry around. If the connection deconstructs and is gone before the transaction does, troubles
```
Comment (utkarshg6) [inline node/src/database/rusqlite_wrappers.rs:98]: Let's omit "can freely".

File: node/src/database/rusqlite_wrappers.rs [hunk 19]
```diff
@@ -43,44 +79,102 @@ impl ConnectionWrapperReal {
...
+// come.
+//
+// The bottom line is that this outer wrapper is designed so that it deconstructs immediately and
+// with its every bit when a commit of the transaction processes.
+
+#[derive(Debug)]
+pub struct TransactionSafeWrapper<'conn_if_real_or_static_if_mock> {
```
Comment (utkarshg6) [inline node/src/database/rusqlite_wrappers.rs:105]: Let's keep the lifetime name to be `'conn`.

Let's add a comment, mentioning that would explain that this lifetime lives for different period of time in real and mock versions.
Real -> Connection
Mock -> Static

**Review by utkarshg6 — COMMENTED** [review]
Continue it from rusqlite_wrappers.rs:109

File: node/src/database/rusqlite_wrappers.rs [hunk 20]
```diff
@@ -43,44 +79,102 @@ impl ConnectionWrapperReal {
...
+
+impl<'conn> TransactionSafeWrapper<'conn> {
+    pub fn new(txn: Transaction<'conn>) -> Self {
+        Self {
+            inner: Box::new(TransactionInnerWrapperReal::new(txn)),
+        }
+    }
+
+    #[cfg(test)]
+    pub fn new_with_builder(inner_wrapper_builder: TransactionInnerWrapperMockBuilder) -> Self {
```
Comment (utkarshg6) [inline node/src/database/rusqlite_wrappers.rs:117]: You and I decided to move this function to a new `impl` block inside `test_utils.rs`. There's another function `arbitrary_id_stamp` that we can also migrate.

File: node/src/database/test_utils/mod.rs [hunk 21]
```diff
@@ -1,28 +1,31 @@
...
-    prepare_results: RefCell<Vec<Result<Statement<'a>, Error>>>,
-    transaction_results: RefCell<Vec<Result<Box<dyn TransactionWrapper + 'a>, Error>>>,
+    prepare_results: RefCell<Vec<Result<Statement<'conn>, Error>>>,
+    transaction_results: RefCell<Vec<Result<TransactionSafeWrapper<'conn>, Error>>>,
 }
 
-unsafe impl<'a> Send for ConnectionWrapperMock<'a> {}
+// We don't know better how to deal with the third-party code for `Statement` that takes on
+// attributes of `Connection` which lacks the Send marker. This unsafe instructs the compiler
+// we don't care because it is a test utility and as far as we know it hasn't bitten us yet
```
Comment (utkarshg6) [inline node/src/database/test_utils/mod.rs:24]: You mentioned, `yet` is redundant because you're using `as far as`.

File: node/src/database/test_utils/mod.rs [hunk 22]
```diff
@@ -1,28 +1,31 @@
...
+pub struct ConnectionWrapperMock<'conn> {
     prepare_params: Arc<Mutex<Vec<String>>>,
-    prepare_results: RefCell<Vec<Result<Statement<'a>, Error>>>,
-    transaction_results: RefCell<Vec<Result<Box<dyn TransactionWrapper + 'a>, Error>>>,
+    prepare_results: RefCell<Vec<Result<Statement<'conn>, Error>>>,
+    transaction_results: RefCell<Vec<Result<TransactionSafeWrapper<'conn>, Error>>>,
 }
 
-unsafe impl<'a> Send for ConnectionWrapperMock<'a> {}
+// We don't know better how to deal with the third-party code for `Statement` that takes on
```
Comment (utkarshg6) [inline node/src/database/test_utils/mod.rs:22]: You also mentioned to change `takes on` to `inherits`.

File: node/src/database/test_utils/transaction_wrapper_mock.rs [hunk 23]
```diff
@@ -2,35 +2,44 @@
...
 use rusqlite::{Error, Statement, ToSql};
 use std::cell::RefCell;
 use std::sync::{Arc, Mutex};
 
-#[derive(Default, Debug)]
-pub struct TransactionWrapperMock {
+// The qualities of this builder are in its own unrestricted usability contrasting with how it
+// creates an encapsulated environment in which it configures and eventually builds the mock. All
+// that in order to minimize the exposure of the mock
+// to the outer scope. While a wrapped transaction is being created, and we need it to be tamed in
```
Comment (utkarshg6) [inline node/src/database/test_utils/transaction_wrapper_mock.rs:18]: You can fix the formatting, if you'd like... not a deal breaker.

File: node/src/database/test_utils/transaction_wrapper_mock.rs [hunk 24]
```diff
@@ -2,35 +2,44 @@
...
+// a test, this mock is supposed to become its inner wrapper of the supposed real world transaction.
```
Comment (utkarshg6) [inline node/src/database/test_utils/transaction_wrapper_mock.rs:19]: this mock is supposed to become the inner of `TransactionSafeWrapper`.

File: node/src/database/test_utils/transaction_wrapper_mock.rs [hunk 25]
```diff
@@ -2,35 +2,44 @@
...
+// The inner wrapper ever needs to stay private. Then there is the outer, public one, that anyone
+// can manipulate with.
+//
+// You can read more about the safeness reasons for this onion structure in comments nearby these
```
Comment (utkarshg6) [inline node/src/database/test_utils/transaction_wrapper_mock.rs:23]: Let's change `onion` -> `layered`.

File: node/src/database/test_utils/transaction_wrapper_mock.rs [hunk 26]
```diff
@@ -2,35 +2,44 @@
...
 
-#[derive(Default, Debug)]
-pub struct TransactionWrapperMock {
+// The qualities of this builder are in its own unrestricted usability contrasting with how it
+// creates an encapsulated environment in which it configures and eventually builds the mock. All
+// that in order to minimize the exposure of the mock
+// to the outer scope. While a wrapped transaction is being created, and we need it to be tamed in
+// a test, this mock is supposed to become its inner wrapper of the supposed real world transaction.
+// The inner wrapper ever needs to stay private. Then there is the outer, public one, that anyone
+// can manipulate with.
```
Comment (utkarshg6) [inline node/src/database/test_utils/transaction_wrapper_mock.rs:21]: `TransactionInnerWrapper` will remain private, but the `TransactionSafeWrapper` will stay public to be used directly.

File: node/src/database/test_utils/transaction_wrapper_mock.rs [hunk 27]
```diff
@@ -44,17 +53,60 @@ impl TransactionWrapperMock {
...
+// Keep as a private class, seek a builder of it instead
+
+#[derive(Debug)]
+struct TransactionInnerWrapperMock {
+    prepare_params: Arc<Mutex<Vec<String>>>,
+    // This field hosts a logical unit able to make the correct
+    // result that seems to be requested to go out based on
+    // the test's setup instructions where the programmer may
+    // have decided to use this advanced test procedure
+    prepare_results_dispatcher_opt: Option<PrepareResultsDispatcher>,
```
Comment (utkarshg6) [inline node/src/database/test_utils/transaction_wrapper_mock.rs:78]: Let's change `dispatcher` to `producer`. Also the type name to `PrepareResultsProducer`.

File: node/src/database/test_utils/transaction_wrapper_mock.rs [hunk 28]
```diff
@@ -44,17 +53,60 @@ impl TransactionWrapperMock {
...
 
-impl TransactionWrapper for TransactionWrapperMock {
+// Keep as a private class, seek a builder of it instead
+
+#[derive(Debug)]
+struct TransactionInnerWrapperMock {
+    prepare_params: Arc<Mutex<Vec<String>>>,
+    // This field hosts a logical unit able to make the correct
+    // result that seems to be requested to go out based on
+    // the test's setup instructions where the programmer may
```
Comment (utkarshg6) [inline node/src/database/test_utils/transaction_wrapper_mock.rs:76]: You may rephrase it to `logical unit able to produce the requested results based on the test's setup instructions`, because we've decided to use the word `produce`.

File: node/src/database/test_utils/transaction_wrapper_mock.rs [hunk 29]
```diff
@@ -44,17 +53,60 @@ impl TransactionWrapperMock {
...
+    ) -> Self {
+        Self {
+            prepare_params,
+            prepare_results_dispatcher_opt,
+            commit_params,
+            commit_results,
+            arbitrary_id_stamp_opt,
+        }
+    }
+}
```
Comment (utkarshg6) [inline node/src/database/test_utils/transaction_wrapper_mock.rs:100]: Let's construct it within the builder, instead of this constructor. Then there won't be a need to provide all the parameters at once.

File: node/src/database/test_utils/transaction_wrapper_mock.rs [hunk 30]
```diff
@@ -66,329 +118,226 @@ impl TransactionWrapper for TransactionWrapperMock {
...
-// the 'prepare' method can have a strong impact on the result returned from the next, upstream
-// function call, to happen on top of this Statement, and which, luckily for us (despite certain
-// difficulties), can be indirectly shaped this way towards our requirements.
-// This distantly prepared Statement can then cause a certainly useful error, easing our test
-// writing. For some examples let's consider the following methods 'execute', 'query_row' or
-// 'query_map'.
+// Trying to store a rusqlite 'Statement' inside the TransactionWrapperMock and then placing this
+// combination into the ConnectionWrapperMock became a thorough test of our Rust knowledge. This
+// approach led to complex problems related to object lifetimes. We tried many times to fix these
+// issues, but success, while never guaranteed, seemed to depend on introduction of numerous new
```
Comment (utkarshg6) [inline node/src/database/test_utils/transaction_wrapper_mock.rs:145]: while __not__ guaranteed.

File: node/src/database/test_utils/transaction_wrapper_mock.rs [hunk 31]
```diff
@@ -66,329 +118,226 @@ impl TransactionWrapper for TransactionWrapperMock {
...
+    // connection meant primarily for the prod-code statements should be also possible.
+    //
+    // Common strategies to use this additional connection:
+    //
+    // a) provide a connection pointing to another database, usually declared
+    //    very simple, that allows a simplified, more direct way of stimulation
+    //    of a special and often quite unusual error that our tested code should
+    //    respond to,
+    //
+    // b) assert on general, unspecific errors while using a connection with
```
Comment (utkarshg6) [inline node/src/database/test_utils/transaction_wrapper_mock.rs:202]: This is a rephrase that we both decided together. It took us several minutes to conclude this.

We can tickle the database while exercising statements that attempt to change the state of the database. Since this connection will be read only, it'll generate an error. Thereby, we can assert on any generic error.

File: node/src/database/test_utils/transaction_wrapper_mock.rs [hunk 32]
```diff
@@ -66,329 +118,226 @@ impl TransactionWrapper for TransactionWrapperMock {
...
+    //    only the read rights for statements whose task is to change the state
+    //    of the database
+    unique_conn_for_altered_stmts_opt: Option<Box<dyn ConnectionWrapper>>,
```
Comment (utkarshg6) [inline node/src/database/test_utils/transaction_wrapper_mock.rs:205]: Maybe rename it to `conn_for_tickling_err_opt`

**Review by utkarshg6 — COMMENTED** [review]
Continue on node/src/database/test_utils/transaction_wrapper_mock.rs:205

File: node/src/database/test_utils/transaction_wrapper_mock.rs [hunk 33]
```diff
@@ -66,329 +118,226 @@ impl TransactionWrapper for TransactionWrapperMock {
...
+    fn commit_prod_code_stmts(&mut self) -> Result<(), Error> {
+        self.txn_bearing_prod_code_stmts_opt
+            .take()
+            .expect("Dual setup with a missing txn should never happen")
+            .commit()
+    }
+
+    fn make_altered_stmt(&self, altered_stm: &str) -> Result<Statement, Error> {
+        match self.unique_conn_for_altered_stmts_opt.as_ref() {
+            None => self
```
Comment (utkarshg6) [inline node/src/database/test_utils/transaction_wrapper_mock.rs:218]: I really liked that you'd like to add a comment mentioning why we'll be using the transaction instead of connection. You mentioned you'd switch the order of `None` and `Some`.

Also, rename the field `unique_conn_for_altered_stmts_opt` to `separate_conn_for_altered_stmts_opt`.

File: node/src/database/test_utils/transaction_wrapper_mock.rs [hunk 34]
```diff
@@ -66,329 +118,226 @@ impl TransactionWrapper for TransactionWrapperMock {
...
+impl Drop for SetupForProdCodeAndAlteredStmts {
     fn drop(&mut self) {
-        // Making sure that the referenced transaction will deconstruct
-        // before the connection it was pointing to
-        drop(self.prod_code_calls_transaction_opt.take())
+        // The real transaction borne in this object binds to a reference that doesn't comply with
+        // safeness anymore, it has gone through backward cast from a raw pointer, which
+        // automatically breaks the check mechanism for referencing an invalid memory. We must make
+        // sure by having this Drop implementation that this transaction deconstructs earlier than
+        // the database Connection by which it was produced, avoiding an OS segmentation error.
```
Comment (utkarshg6) [inline node/src/database/test_utils/transaction_wrapper_mock.rs:234]: maybe add it like this... `database Connection (which is part of the struct)`

File: node/src/database/test_utils/transaction_wrapper_mock.rs [hunk 35]
```diff
@@ -66,329 +118,226 @@ impl TransactionWrapper for TransactionWrapperMock {
...
+        drop(self.txn_bearing_prod_code_stmts_opt.take())
     }
 }
 
 #[derive(Debug)]
-pub struct PrepareMethodResults {
-    calls_counter: RefCell<usize>,
-    setup: Either<SetupForBoth, SetupForStubbed<String>>,
+pub struct PrepareResultsDispatcher {
```
Comment (utkarshg6) [inline node/src/database/test_utils/transaction_wrapper_mock.rs:240]: I've already left another comment to rename this, in case the rename was missed during that, you may do it using this comment.

File: node/src/database/test_utils/transaction_wrapper_mock.rs [hunk 36]
```diff
@@ -66,329 +118,226 @@ impl TransactionWrapper for TransactionWrapperMock {
...
+    setup: Either<SetupForOnlyAlteredStmts, SetupForProdCodeAndAlteredStmts>,
 }
 
-impl PrepareMethodResults {
-    pub fn new_with_both_prod_code_and_stubbed_calls(
-        prod_code_calls_conn: Box<dyn ConnectionWrapper>,
-        stubbed_calls_conn: Box<dyn ConnectionWrapper>,
+impl PrepareResultsDispatcher {
+    pub fn new_with_altered_stmts_only(
```
Comment (utkarshg6) [inline node/src/database/test_utils/transaction_wrapper_mock.rs:245]: We decided that we can rename it to `construct_with_altered_stmts_only`. Although, you find it embarrassing that it's long, I don't find it embarrassing at all. I find it clear and explicit.

File: node/src/database/test_utils/transaction_wrapper_mock.rs [hunk 37]
```diff
@@ -66,329 +118,226 @@ impl TransactionWrapper for TransactionWrapperMock {
...
+// in the test. This approach can be applied to various methods such as 'execute', 'query_row',
+// or 'query_map'.
 
 #[derive(Debug)]
-struct SetupForStubbed<S> {
-    stubbed_calls_conn: Box<dyn ConnectionWrapper>,
-    stubbed_calls_optional_statements_literals: Vec<S>,
+struct SetupForOnlyAlteredStmts {
+    conn: Box<dyn ConnectionWrapper>,
+    queue_of_statements: RefCell<Vec<AlteredStmtByOrigin>>,
```
Comment (utkarshg6) [inline node/src/database/test_utils/transaction_wrapper_mock.rs:178]: You mentioned that we can make it simpler, if we just use the `String` here instead of using `AtererdStmtByOrigin`.

File: node/src/database/test_utils/transaction_wrapper_mock.rs [hunk 38]
```diff
@@ -66,329 +118,226 @@ impl TransactionWrapper for TransactionWrapperMock {
...
+        prod_code_stmt: &str,
+    ) -> Result<Statement<'conn>, Error> {
+        let altered_stmt_opt = setup
+            .queue_with_prod_code_and_altered_stmts
+            .borrow_mut()
+            .remove(0);
+
+        match altered_stmt_opt {
+            None => setup.prod_code_stmts_conn.prepare(prod_code_stmt),
+            Some(altered_stm) => {
```
Comment (utkarshg6) [inline node/src/database/test_utils/transaction_wrapper_mock.rs:317]: Let's change it to `aletered_stmt`.

File: node/src/database/test_utils/transaction_wrapper_mock.rs [hunk 39]
```diff
@@ -66,329 +118,226 @@ impl TransactionWrapper for TransactionWrapperMock {
...
+    IdenticalWithProdCode,
+    FromSubstitution { new_stmt: String },
 }
 
-impl StubbedCallIndexInfo {
-    fn new(preceding_prod_code_calls: usize) -> Self {
-        Self {
-            preceding_prod_code_calls,
+impl AlteredStmtByOrigin {
+    fn str_stmt<'a>(&'a self, prod_code_stm: &'a str) -> &'a str {
```
Comment (utkarshg6) [inline node/src/database/test_utils/transaction_wrapper_mock.rs:335]: I have this concern:

If a statement is passed in `prod_code_stm`, it'll be expected that this function will make a use of it. In some cases, it'll pass on the `FromSubstitution` arm, and it won't get used.

The problem is that the programmer will not even realize that the statement wasn't used.

Any help is appreciated. :+1:  :100:

File: node/src/db_config/config_dao.rs [hunk 40]
```diff
@@ -91,13 +91,16 @@ impl ConfigDaoReal {
...
+    fn set_value<'a, P>(
+        prepare_stm: P,
+        name: &str,
+        value: Option<String>,
+    ) -> Result<(), ConfigDaoError>
     where
         P: FnOnce(&str) -> Result<Statement<'a>, rusqlite::Error> + 'a,
     {
-        let mut stmt = match prepare("update config set value = ? where name = ?") {
+        let mut stmt = match prepare_stm("update config set value = ? where name = ?") {
```
Comment (utkarshg6) [inline node/src/db_config/config_dao.rs:102]: Let's change it to the `prepare_stmt`.

File: node/src/db_config/config_dao.rs [hunk 41]
```diff
@@ -287,16 +291,23 @@ mod tests {
...
-    fn setting_by_started_transaction_works() {
+    fn set_by_guest_transaction_works() {
         let home_dir =
-            ensure_node_home_directory_exists("config_dao", "setting_by_started_transaction_works");
-        // To make the test even more convincing, let's create a useless "Connection".
+            ensure_node_home_directory_exists("config_dao", "set_by_guest_transaction_works");
+        // To make the test even more convincing, let's give a useless "Connection" to the hosting DAO object,
+        // while the guest connection will do the job
         let subject = ConfigDaoReal::new(Box::new(ConnectionWrapperMock::default()));
         let mut conn = initialize_conn(&home_dir);
```
Comment (utkarshg6) [inline node/src/db_config/config_dao.rs:300]: It creates the new database with the version number according to the constant `CURRENT_SCHEMA_VERSION`. You wanted to make it a part of this test as a comment.

**Review by utkarshg6 — COMMENTED** [review]
Continue at node/src/db_config/config_dao.rs:300

**Review by utkarshg6 — COMMENTED** [review]
There were really nice changes done!

I think after this review, it would be even nicer (:
